core/src/main/java/org/elasticsearch/plugins/SearchPlugin.java

```diff
@@ -0,0 +1,173 @@
+import org.elasticsearch.search.aggregations.bucket.significant.heuristics.SignificanceHeuristicParser;
+import org.elasticsearch.search.aggregations.pipeline.movavg.models.MovAvgModel;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.highlight.Highlighter;
+import org.elasticsearch.search.suggest.Suggester;
+
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+
+/**
+ * Plugin for extending search time behavior.
+ */
+public interface SearchPlugin {
+    /**
+     * The new {@link ScoreFunction}s defined by this plugin.
+     */
+    default List<ScoreFunctionSpec<?>> getScoreFunctions() {
+        return emptyList();
+    }
+    /**
+     * The new {@link SignificanceHeuristic}s defined by this plugin.
+     */
+    default List<SearchPluginSpec<SignificanceHeuristic, SignificanceHeuristicParser>> getSignificanceHeuristics() {
+        return emptyList();
+    }
+    /**
+     * The new {@link MovAvgModel}s defined by this plugin.
+     */
+    default List<SearchPluginSpec<MovAvgModel, MovAvgModel.AbstractModelParser>> getMovingAverageModels() {
+        return emptyList();
+    }
+    /**
+     * The new {@link FetchSubPhase}s defined by this plugin.
+     */
+    default List<FetchSubPhase> getFetchSubPhases(FetchPhaseConstructionContext context) {
+        return emptyList();
+    }
+    /**
+     * Get the {@link Highlighter}s defined by this plugin.
+     */
+    default Map<String, Highlighter> getHighlighters() {
+        return emptyMap();
+    }
+    /**
+     * The new {@link Suggester}s defined by this plugin.
+     */
+    default Map<String, Suggester<?>> getSuggesters() {
+        return emptyMap();
+    }
+
+    /**
+     * Specification of custom {@link ScoreFunction}.
+     */
+    public class ScoreFunctionSpec<T extends ScoreFunctionBuilder<T>> extends SearchPluginSpec<T, ScoreFunctionParser<T>> {
+        public ScoreFunctionSpec(ParseField name, Reader<T> reader, ScoreFunctionParser<T> parser) {
+            super(name, reader, parser);
+        }
+
+        public ScoreFunctionSpec(String name, Reader<T> reader, ScoreFunctionParser<T> parser) {
+            super(name, reader, parser);
+        }
+    }
+
+    /**
+     * Specification of search time behavior extension like a custom {@link MovAvgModel} or {@link ScoreFunction}.
+     *
+     * @param W the type of the main {@link NamedWriteable} for this spec. All specs have this but it isn't always *for* the same thing
+     *        though, usually it is some sort of builder sent from the coordinating node to the data nodes executing the behavior
+     * @param P the type of the parser for this spec. The parser runs on the coordinating node, converting {@link XContent} into the
+     *        behavior to execute
+     */
+    public class SearchPluginSpec<W extends NamedWriteable, P> {
```

Comment:
Does this really need to be generic? We certainly don't care about any of that on the consumption side.